src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp

```diff
@@ -0,0 +1,554 @@
+      blockRequests[ i ] = blockRequest;
+      QObject::connect( blockRequest, &QgsPointCloudBlockRequest::finished, &loop, [ &, i, nStr, blockRequest ]()
+      {
+        if ( !blockRequest->block() )
+        {
+          QgsDebugMsg( QStringLiteral( "Unable to load node %1, error: %2" ).arg( nStr, blockRequest->errorStr() ) );
+        }
+        finishedLoadingBlock[ i ] = true;
+        // If all blocks are loaded, exit the event loop
+        if ( !finishedLoadingBlock.contains( false ) ) loop.exit();
+      } );
+    }
+    // Wait for all point cloud nodes to finish loading
+    loop.exec();
+
+    if ( !mFeedback->isCanceled() )
+    {
+      // Render all the point cloud blocks sequentially
+      for ( int i = 0; i < blockRequests.size(); ++i )
+      {
+        if ( mFeedback->isCanceled() )
+        {
+          break;
+        }
+
+        if ( !blockRequests[ i ]->block() )
+          continue;
+
+        visitBlock( blockRequests[ i ]->block() );
+
+        ++nodesDrawn;
+      }
+    }
+
+    for ( int i = 0; i < blockRequests.size(); ++i )
+    {
+      if ( blockRequests[ i ] )
+      {
+        if ( blockRequests[ i ]->block() )
+          delete blockRequests[ i ]->block();
+        blockRequests[ i ]->deleteLater();
+      }
+    }
+  }
+
+  return nodesDrawn;
+}
+
+void QgsPointCloudLayerProfileGenerator::visitBlock( const QgsPointCloudBlock *block )
+{
+  const char *ptr = block->data();
+  int count = block->pointCount();
+
+  const QgsPointCloudAttributeCollection request = block->attributes();
+
+  const std::size_t recordSize = request.pointRecordSize();
+
+  const QgsPointCloudAttributeCollection blockAttributes = block->attributes();
+  int xOffset = 0, yOffset = 0, zOffset = 0;
+  const QgsPointCloudAttribute::DataType xType = blockAttributes.find( QStringLiteral( "X" ), xOffset )->type();
+  const QgsPointCloudAttribute::DataType yType = blockAttributes.find( QStringLiteral( "Y" ), yOffset )->type();
+  const QgsPointCloudAttribute::DataType zType = blockAttributes.find( QStringLiteral( "Z" ), zOffset )->type();
+
+  const bool reproject = !mLayerToTargetTransform.isShortCircuited();
+  for ( int i = 0; i < count; ++i )
+  {
+    if ( mFeedback->isCanceled() )
+    {
+      break;
+    }
+
+    QgsPointCloudLayerProfileResults::PointResult res;
+    QgsPointCloudAttribute::getPointXYZ( ptr, i, recordSize, xOffset, xType, yOffset, yType, zOffset, zType, block->scale(), block->offset(), res.x, res.y, res.z );
+
+    res.z = res.z * mZScale + mZOffset;
+
+    if ( mSearchGeometryInLayerCrsGeometryEngine->contains( res.x, res.y ) )
+    {
+      if ( reproject )
+      {
+        try
+        {
+          mLayerToTargetTransform.transformInPlace( res.x, res.y, res.z );
+        }
+        catch ( QgsCsException & )
+        {
+          continue;
+        }
+      }
+
+      res.color = mPointColor.rgba();
+
+      // TODO we may hit the limit of QVector size?
```

Comment:
I hope that should be unlikely if user didn't pick extremely big tolerance for the curve and extremely low max. error... (for full-canvas rendering of point clouds we typically draw up few millions of points)